apps/prairielearn/src/lib/authz-data-lib.ts

```diff
@@ -0,0 +1,183 @@
+  authn_has_student_access_with_enrollment: z.boolean().optional(),
+  authn_has_course_instance_permission_view: z.boolean().optional(),
+  authn_has_course_instance_permission_edit: z.boolean().optional(),
+
+  // Authz data
+  user: RawStaffUserSchema,
+  is_administrator: z.boolean(),
+  has_course_permission_preview: z.boolean(),
+  has_course_permission_view: z.boolean(),
+  has_course_permission_edit: z.boolean(),
+  has_course_permission_own: z.boolean(),
+  course_role: EnumCourseRoleSchema.optional(),
+  course_instance_role: EnumCourseInstanceRoleSchema.optional(),
+  mode: EnumModeSchema.optional(),
+  has_student_access: z.boolean().optional(),
+  has_student_access_with_enrollment: z.boolean().optional(),
+  has_course_instance_permission_view: z.boolean().optional(),
+  has_course_instance_permission_edit: z.boolean().optional(),
+});
+export type RawPageAuthzData = z.infer<typeof RawPageAuthzDataSchema>;
+
+export const PageAuthzDataSchema = RawPageAuthzDataSchema.extend({
+  user: StaffUserSchema,
+  authn_user: StaffUserSchema,
+}).brand<'PageAuthzData'>();
+export type PageAuthzData = z.infer<typeof PageAuthzDataSchema>;
+
+export interface DangerousSystemAuthzData {
+  authn_user: {
+    user_id: null;
+  };
+  user: {
+    user_id: null;
+  };
+}
+
+/** The full authz data from a database query. This is NOT what is on res.locals. */
+export const FullAuthzDataSchema = z.object({
+  mode: EnumModeSchema,
+  mode_reason: EnumModeReasonSchema,
+  course: CourseSchema,
+  institution: InstitutionSchema,
+  course_instance: CourseInstanceSchema.nullable(),
+  permissions_course: SprocAuthzCourseSchema,
+  permissions_course_instance: SprocAuthzCourseInstanceSchema,
+});
+
+/** The user facing version that is obtained via the page context helpers. */
+export type AuthzData = RawPageAuthzData | DangerousSystemAuthzData;
+
+export type CourseInstanceRole =
+  | 'None'
+  | 'Student'
+  | 'Student Data Viewer'
+  | 'Student Data Editor'
+  // The role 'Any' is equivalent to 'Student' OR 'Student Data Viewer' OR 'Student Data Editor'
+  | 'Any';
+
+export function dangerousFullAuthzForTesting(): DangerousSystemAuthzData {
+  return {
+    authn_user: {
+      // We use this structure with a user_id of null to indicate that the user is the system.
+      // Inserts into the audit_events table as a system user have a user_id of null.
+      user_id: null,
+    },
+    user: {
+      user_id: null,
+    },
+  };
+}
+
+export function isDangerousFullAuthzForTesting(
+  authzData: AuthzData,
+): authzData is DangerousSystemAuthzData {
+  return authzData.authn_user.user_id === null && authzData.user.user_id === null;
+}
+
+export function hasRole(authzData: AuthzData, requestedRole: CourseInstanceRole): boolean {
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  if (!authzData) {
+    throw new Error('authzData is not defined');
+  }
+
+  if (isDangerousFullAuthzForTesting(authzData)) {
```

Comment:
One of the PR changes the semantics of this to only accept `System` or `Any` here, good job not pulling that in.